### PR TITLE
[Core] use higher niceness for workers

### LINKED ISSUE #25012: [Core] Received a GOAWAY with error code ENHANCE_YOUR_CALM and debug data equal to "too_many_pings"
### What happened + What you expected to happen

I got an error when running the following code with **multi-worker**:
ERROR:
```
(RolloutWorker pid=2591) E0520 15:17:56.430216750    2847 chttp2_transport.cc:1111]   **Received a GOAWAY with error code ENHANCE_YOUR_CALM and debug data equal to "too_many_pings"**
(RolloutWorker pid=2592) E0520 15:17:56.429997321    2815 chttp2_transport.cc:1111]   Received a GOAWAY with error code ENHANCE_YOUR_CALM and debug data equal to "too_many_pings"
(RolloutWorker pid=2588) E0520 15:17:56.429843491    2814 chttp2_transport.cc:1111]   Received a GOAWAY with error code ENHANCE_YOUR_CALM and debug data equal to "too_many_pings"
(RolloutWorker pid=2590) E0520 15:17:56.430071007    2816 chttp2_transport.cc:1111]   Received a GOAWAY with error code ENHANCE_YOUR_CALM and debug data equal to "too_many_pings"
(RolloutWorker pid=2589) E0520 15:17:56.429672231    2782 chttp2_transport.cc:1111]   Received a GOAWAY with error code ENHANCE_YOUR_CALM and debug data equal to "too_many_pings"
(RolloutWorker pid=2589) 2022-05-20 15:17:56,437	ERROR worker.py:449 -- SystemExit was raised from the worker.
(RolloutWorker pid=2589) Traceback (most recent call last):
(RolloutWorker pid=2589)   File "python/ray/_raylet.pyx", line 797, in ray._raylet.task_execution_handler
(RolloutWorker pid=2589)   File "python/ray/_raylet.pyx", line 616, in ray._raylet.execute_task
(RolloutWorker pid=2589)   File "python/ray/_raylet.pyx", line 656, in ray._raylet.execute_task
(RolloutWorker pid=2589)   File "python/ray/_raylet.pyx", line 663, in ray._raylet.execute_task
(RolloutWorker pid=2589)   File "python/ray/_raylet.pyx", line 667, in ray._raylet.execute_task
(RolloutWorker pid=2589)   File "python/ray/_raylet.pyx", line 614, in ray._raylet.execute_task.function_executor
(RolloutWorker pid=2589)   File "/home/mhx/anaconda3/envs/RLlib/lib/python3.8/site-packages/ray/_private/function_manager.py", line 701, in actor_method_executor
(RolloutWorker pid=2589)     return method(__ray_actor, *args, **kwargs)
(RolloutWorker pid=2589)   File "/home/mhx/anaconda3/envs/RLlib/lib/python3.8/site-packages/ray/util/tracing/tracing_helper.py", line 462, in _resume_span
(RolloutWorker pid=2589)     return method(self, *_args, **_kwargs)
(RolloutWorker pid=2589)   File "/home/mhx/anaconda3/envs/RLlib/lib/python3.8/site-packages/ray/actor.py", line 1324, in __ray_terminate__
(RolloutWorker pid=2589)     ray.actor.exit_actor()
(RolloutWorker pid=2589)   File "/home/mhx/anaconda3/envs/RLlib/lib/python3.8/site-packages/ray/actor.py", line 1412, in exit_actor
(RolloutWorker pid=2589)     ray.worker.disconnect()
(RolloutWorker pid=2589)   File "/home/mhx/anaconda3/envs/RLlib/lib/python3.8/site-packages/ray/worker.py", line 1672, in disconnect
(RolloutWorker pid=2589)     worker.gcs_error_subscriber.close()
(RolloutWorker pid=2589)   File "/home/mhx/anaconda3/envs/RLlib/lib/python3.8/site-packages/ray/_private/gcs_pubsub.py", line 306, in close
(RolloutWorker pid=2589)     self._stub.GcsSubscriberCommandBatch(req, timeout=5)
(RolloutWorker pid=2589)   File "/home/mhx/anaconda3/envs/RLlib/lib/python3.8/site-packages/grpc/_channel.py", line 944, in __call__
(RolloutWorker pid=2589)     state, call, = self._blocking(request, timeout, metadata, credentials,
(RolloutWorker pid=2589)   File "/home/mhx/anaconda3/envs/RLlib/lib/python3.8/site-packages/grpc/_channel.py", line 933, in _blocking
(RolloutWorker pid=2589)     event = call.next_event()
(RolloutWorker pid=2589)   File "src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi", line 338, in grpc._cython.cygrpc.SegregatedCall.next_event
(RolloutWorker pid=2589)   File "src/python/grpcio/grpc/_cython/_cygrpc/channel.pyx.pxi", line 171, in grpc._cython.cygrpc._next_call_event
(RolloutWorker pid=2589)   File "/home/mhx/anaconda3/envs/RLlib/lib/python3.8/threading.py", line 246, in __enter__
(RolloutWorker pid=2589)     def __enter__(self):
(RolloutWorker pid=2589)   File "/home/mhx/anaconda3/envs/RLlib/lib/python3.8/site-packages/ray/worker.py", line 446, in sigterm_handler
(RolloutWorker pid=2589)     sys.exit(1)
(RolloutWorker pid=2589) SystemExit: 1
(RolloutWorker pid=2587) E0520 15:17:56.429773197    2750 chttp2_transport.cc:1111]   Received a GOAWAY with error code ENHANCE_YOUR_CALM and debug data equal to "too_many_pings"
(RolloutWorker pid=2596) E0520 15:17:56.430210222    2848 chttp2_transport.cc:1111]   Received a GOAWAY with error code ENHANCE_YOUR_CALM and debug data equal to "too_many_pings"
(RolloutWorker pid=2602) E0520 15:17:56.430515333    2867 chttp2_transport.cc:1111]   Received a GOAWAY with error code ENHANCE_YOUR_CALM and debug data equal to "too_many_pings"
(RolloutWorker pid=2601) E0520 15:17:56.430439596    2868 chttp2_transport.cc:1111]   Received a GOAWAY with error code ENHANCE_YOUR_CALM and debug data equal to "too_many_pings"
2022-05-20 15:17:56,547	INFO tune.py:701 -- Total run time: 250.95 seconds (250.69 seconds for the tuning loop).
```

### Versions / Dependencies

python 3.8

Package                  Version
------------------------ ---------
aiohttp                  3.8.1
aiohttp-cors             0.7.0
aioredis                 1.3.1
aiorwlock                1.3.0
aiosignal                1.2.0
anyio                    3.6.1
asgiref                  3.5.2
async-timeout            4.0.1
attrs                    21.2.0
blessed                  1.19.0
cachetools               4.2.4
certifi                  2021.10.8
charset-normalizer       2.0.7
click                    8.0.3
cloudpickle              1.6.0
colorful                 0.5.4
cycler                   0.11.0
Deprecated               1.2.13
dgl                      0.7.2
distlib                  0.3.4
dm-tree                  0.1.6
docutils                 0.18.1
fastapi                  0.78.0
filelock                 3.4.0
fonttools                4.28.2
frozenlist               1.2.0
future                   0.18.2
google-api-core          2.2.2
google-auth              2.3.3
googleapis-common-protos 1.53.0
gpustat                  1.0.0b1
grpcio                   1.42.0
gym                      0.18.0
h11                      0.13.0
hiredis                  2.0.0
idna                     3.3
imageio                  2.9.0
importlib-resources      5.4.0
jsonschema               4.2.1
kiwisolver               1.3.2
lz4                      3.1.3
matplotlib               3.5.0
mkl-fft                  1.3.1
mkl-random               1.2.2
mkl-service              2.4.0
msgpack                  1.0.2
multidict                5.2.0
networkx                 2.6.3
numpy                    1.21.4
nvidia-ml-py3            7.352.0
olefile                  0.46
opencensus               0.8.0
opencensus-context       0.1.2
packaging                21.3
pandas                   1.3.4
Pillow                   7.2.0
pip                      21.2.4
platformdirs             2.5.2
prometheus-client        0.12.0
protobuf                 3.19.1
psutil                   5.8.0
py-spy                   0.3.11
pyasn1                   0.4.8
pyasn1-modules           0.2.8
pydantic                 1.9.1
pyglet                   1.5.0
pyparsing                3.0.6
pyrsistent               0.18.0
python-dateutil          2.8.2
pytz                     2021.3
PyWavelets               1.2.0
PyYAML                   6.0
ray                      1.12.1
redis                    4.0.2
requests                 2.26.0
rsa                      4.7.2
scikit-image             0.18.3
scipy                    1.7.1
setuptools               58.0.4
setuptools-scm           6.3.2
six                      1.16.0
smart-open               6.0.0
sniffio                  1.2.0
starlette                0.19.1
tabulate                 0.8.9
tensorboardX             2.4.1
tifffile                 2021.11.2
tomli                    1.2.2
torch                    1.10.0
torchaudio               0.10.0
torchvision              0.11.1
typing-extensions        3.10.0.2
urllib3                  1.26.7
uvicorn                  0.16.0
virtualenv               20.14.1
wcwidth                  0.2.5
wheel                    0.37.0
wrapt                    1.13.3
yarl                     1.7.2
zipp                     3.6.0

### Reproduction script
```
import ray
from ray import tune
from ray.rllib.agents.dqn import DQNTrainer
ray.shutdown()
ray.init()  
tune.run(DQNTrainer, config={"env": "CartPole-v0", 'framework': 'torch',"train_batch_size": 400,  'num_workers': 10, 'output': '/tmp/cartpole-out', 'output_max_file_size': 100}, checkpoint_at_end=True, checkpoint_freq=10, stop={'timesteps_total': 100000})
```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
Looking at past failures of `dataset_shuffle_push_based_random_shuffle_1tb` and when running it on my own, I noticed that raylets are killed because GCS was not able to respond to it in time. It seems at the beginning of the run, there is a huge CPU spike which starved GCS out of CPU. With the same spirit of adjusting workers to higher OOM scores, we can give workers higher niceness so they yield CPU to GCS, Raylet and other user processes.

I ran `dataset_shuffle_push_based_random_shuffle_1tb` a few time which no longer sees raylet death because of GCS CPU starvation. But there are other issues making the test fail which I will continue to investigate.

## Related issue number
Closes #25012

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_placement_group_mini_integration.py [hunk 1]
```diff
@@ -23,7 +23,7 @@ def run_mini_integration_test(cluster, pg_removal=True, num_pgs=999):
     # When 999 resource quantity is used, it fails about every other time
     # when the test was written.
     resource_quantity = num_pgs
-    num_nodes = 5
+    num_nodes = 4
```
Comment (stephanie-wang) [inline python/ray/tests/test_placement_group_mini_integration.py:26]: What's the reason for this change?

File: src/ray/raylet/worker_pool.cc [hunk 2]
```diff
@@ -364,6 +374,34 @@ std::tuple<Process, StartupToken> WorkerPool::StartWorkerProcess(
...
+    } else if (language == Language::JAVA) {
+      worker_command_args.push_back("--language=" + Language_Name(language));
+    }
+
+    if (ray_debugger_external) {
+      worker_command_args.push_back("--ray-debugger-external");
+    }
+  }
+
+  if (const std::vector<std::string> args = CommandArgsToAdjustWorkerNice();
```
Comment (stephanie-wang) [inline src/ray/raylet/worker_pool.cc:400]: Can we instead call nice directly?

The [man page](https://linux.die.net/man/3/nice) has an example.

**Review by stephanie-wang — CHANGES_REQUESTED** [review]

**Comment by stephanie-wang** [comment]
"Nice" find! :)

File: python/ray/tests/test_placement_group_mini_integration.py (hunk 1 above)
Comment (mwtian) [reply python/ray/tests/test_placement_group_mini_integration.py:26]: It seems because unit test VMs are very limited in CPU, lowering CPU priority of workers resulted in some workers unable to register with Raylet. I'm trying to see if reducing the number of workers somewhat can fix the issue. And I will try calling `nice()` after coreworker registers with raylet.

Comment (stephanie-wang) [reply python/ray/tests/test_placement_group_mini_integration.py:26]: Got it, thanks!

File: src/ray/raylet/worker_pool.cc [hunk 3]
```diff
@@ -364,6 +362,29 @@ std::tuple<Process, StartupToken> WorkerPool::StartWorkerProcess(
                                   std::to_string(worker_startup_token_counter_));
   }
 
+  if (language == Language::PYTHON || language == Language::JAVA) {
```
Comment (mwtian) [inline src/ray/raylet/worker_pool.cc:365]: This just moves the block of code so `worker_command_args` handling logic are located together. Although this is unrelated to updating niceness, I prefer to keep this cleanup.

File: src/ray/core_worker/core_worker.cc [hunk 4]
```diff
@@ -516,6 +520,13 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
         }
       },
       100);
+
+#ifndef _WIN32
+  // nice() increments the niceness of this process, so this assumes one CoreWorker object
+  // per process.
+  const auto niceness = nice(RayConfig::instance().worker_niceness());
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc:527]: Please add a note that we should only do this after the worker has registered to the raylet (otherwise we could thrash trying to start workers when under pressure).

**Review by stephanie-wang — APPROVED** [review]

File: src/ray/common/ray_config_def.h [hunk 5]
```diff
@@ -242,6 +242,13 @@ RAY_CONFIG(uint32_t, worker_max_resource_analysis_iteration, 128);
 /// NOTE(swang): Linux only.
 RAY_CONFIG(int, worker_oom_score_adjustment, 1000)
 
+/// Sets workers' nice value on posix systems, so that the OS prioritizes CPU for other
+/// processes over worker. This makes CPU available to GCS, Raylet and user processes
+/// even when workers are busy.
+/// Valid value is [0, 19] (negative values require sudo permissions).
+/// NOTE: Linux, Unix and MacOS only.
+RAY_CONFIG(int, worker_niceness, 15)
```
Comment (rkooo567) [inline src/ray/common/ray_config_def.h:250]: Is the default niceness different from platform & OS?

Comment (rkooo567) [reply src/ray/common/ray_config_def.h:250]: and I wonder in this case, if we should just set low niceness on raylet instead of increasing worker niceness (e..g, set the lowest possible niceness on raylet)?

Comment (rkooo567) [reply src/ray/common/ray_config_def.h:250]: since it means worker process can get less CPU than other processes like e.g., log moniitor

File: src/ray/core_worker/core_worker.cc [hunk 6]
```diff
@@ -516,6 +520,21 @@ CoreWorker::CoreWorker(const CoreWorkerOptions &options, const WorkerID &worker_
...
+  static const bool niced = [this]() {
+    if (options_.worker_type != WorkerType::DRIVER) {
+      const auto niceness = nice(RayConfig::instance().worker_niceness());
+      RAY_LOG(INFO) << "Adjusted worker niceness to " << niceness;
+      return true;
+    }
+    return false;
+  }();
+  // Verify driver and worker are never mixed in the same process.
+  RAY_CHECK_EQ(options_.worker_type != WorkerType::DRIVER, niced);
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:536]: is there any benefit of doing this in this way compared to
```
if (worker_Type != DRIVER) {
      const auto niceness = nice(RayConfig::instance().worker_niceness());
      RAY_LOG(INFO) << "Adjusted worker niceness to " << niceness;
}
```
?

File: src/ray/common/ray_config_def.h (hunk 5 above)
Comment (mwtian) [reply src/ray/common/ray_config_def.h:250]: Lowering niceness on Raylet requires `sudo`. Also, having processes like reporter agent, dashboard with lower niceness is desirable because they are required for debugging. IIUC log monitor is not actually related to log persistence, so if it turns out to consume huge amount of CPU we can optimize that and maybe give it a higher niceness.

File: src/ray/core_worker/core_worker.cc (hunk 6 above)
Comment (mwtian) [reply src/ray/core_worker/core_worker.cc:536]: Just want to make it clear that we want to increment niceness only once per-process, and process that updates niceness cannot be the driver.
